File: android/app/src/main/java/com/studyapp/presentation/plans/PlanViewModel.kt
package com.studyapp.presentation.plans

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.studyapp.domain.model.*
import com.studyapp.domain.repository.SubjectRepository
import com.studyapp.domain.usecase.ManagePlansUseCase
import com.studyapp.domain.util.Result
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import java.time.DayOfWeek
import javax.inject.Inject

data class PlanUiState(
    val activePlan: StudyPlan? = null,
    val planItems: List<PlanItem> = emptyList(),
    val subjects: List<Subject> = emptyList(),
    val weeklySchedule: Map<DayOfWeek, List<PlanItemWithSubject>> = emptyMap(),
    val totalTargetMinutes: Int = 0,
    val totalActualMinutes: Int = 0,
    val isLoading: Boolean = true,
    val error: String? = null
)

@HiltViewModel
class PlanViewModel @Inject constructor(
    private val managePlansUseCase: ManagePlansUseCase,
    private val subjectRepository: SubjectRepository
) : ViewModel() {
    
    private val _uiState = MutableStateFlow(PlanUiState())
    val uiState: StateFlow<PlanUiState> = _uiState.asStateFlow()
    
    init {
        observeData()
    }
    
    private fun observeData() {
        viewModelScope.launch {
            subjectRepository.getAllSubjects()
                .combine(managePlansUseCase.getActivePlan()) { subjectsResult, plan ->
                    val subjects = when (subjectsResult) {
                        is Result.Success -> subjectsResult.data
                        is Result.Error -> {
                            _uiState.update { it.copy(
                                isLoading = false,
                                error = subjectsResult.message ?: subjectsResult.exception.message
                            )}
                            return@combine null
                        }
                    }
                    Pair(subjects, plan)
                }
                .filterNotNull()
                .flatMapLatest { (subjects, plan) ->
                    if (plan != null) {
                        managePlansUseCase.getPlanItems(plan.id).map { items ->
                            Triple(subjects, plan, items)
                        }
                    } else {
                        flowOf(Triple(subjects, null, emptyList()))
                    }
                }
                .collect { (subjects, plan, items) ->
                    val weeklySchedule = buildWeeklySchedule(items, subjects)
                    val totalTarget = items.sumOf { it.targetMinutes }
                    
                    _uiState.update { state ->
                        state.copy(
                            activePlan = plan,
                            planItems = items,
                            subjects = subjects,
                            weeklySchedule = weeklySchedule,
                            totalTargetMinutes = totalTarget,
                            isLoading = false,
                            error = null
                        )
                    }
                }
        }
    }
    
    private fun buildWeeklySchedule(
        items: List<PlanItem>,
        subjects: List<Subject>
    ): Map<DayOfWeek, List<PlanItemWithSubject>> {
        val subjectMap = subjects.associateBy { it.id }
        
        return DayOfWeek.entries.associateWith { day ->
            items
                .filter { it.dayOfWeek == day }
                .mapNotNull { item ->
                    subjectMap[item.subjectId]?.let { subject ->
                        PlanItemWithSubject(item, subject)
                    }
                }
        }
    }
    
    fun createPlan(name: String, startDate: Long, endDate: Long, items: List<PlanItem>) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            
            val plan = StudyPlan(
                name = name,
                startDate = startDate,
                endDate = endDate,
                createdAt = System.currentTimeMillis()
            )
            
            when (val result = managePlansUseCase.createPlan(plan, items)) {
                is Result.Success -> {
                    _uiState.update { it.copy(isLoading = false, error = null) }
                }
                is Result.Error -> {
                    _uiState.update { it.copy(
                        isLoading = false,
                        error = result.message ?: result.exception.message
                    )}
                }
            }
        }
    }
    
    fun addPlanItem(subjectId: Long, dayOfWeek: DayOfWeek, targetMinutes: Int, timeSlot: String?) {
        viewModelScope.launch {
            val plan = _uiState.value.activePlan ?: return@launch
            
            _uiState.update { it.copy(isLoading = true) }
            
            val item = PlanItem(
                planId = plan.id,
                subjectId = subjectId,
                dayOfWeek = dayOfWeek,
                targetMinutes = targetMinutes,
                timeSlot = timeSlot
            )
            
            when (val result = managePlansUseCase.addPlanItem(item)) {
                is Result.Success -> {
                    _uiState.update { it.copy(isLoading = false, error = null) }
                }
                is Result.Error -> {
                    _uiState.update { it.copy(
                        isLoading = false,
                        error = result.message ?: result.exception.message
                    )}
                }
            }
        }
    }
    
    fun updatePlanItem(item: PlanItem) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            
            when (val result = managePlansUseCase.updatePlanItem(item)) {
                is Result.Success -> {
                    _uiState.update { it.copy(isLoading = false, error = null) }
                }
                is Result.Error -> {
                    _uiState.update { it.copy(
                        isLoading = false,
                        error = result.message ?: result.exception.message
                    )}
                }
            }
        }
    }
    
    fun deletePlanItem(item: PlanItem) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            
            when (val result = managePlansUseCase.deletePlanItem(item)) {
                is Result.Success -> {
                    _uiState.update { it.copy(isLoading = false, error = null) }
                }
                is Result.Error -> {
                    _uiState.update { it.copy(
                        isLoading = false,
                        error = result.message ?: result.exception.message
                    )}
                }
            }
        }
    }
    
    fun deletePlan() {
        viewModelScope.launch {
            val plan = _uiState.value.activePlan ?: return@launch
            
            _uiState.update { it.copy(isLoading = true) }
            
            when (val result = managePlansUseCase.deletePlan(plan)) {
                is Result.Success -> {
                    _uiState.update { it.copy(isLoading = false, error = null) }
                }
                is Result.Error -> {
                    _uiState.update { it.copy(
                        isLoading = false,
                        error = result.message ?: result.exception.message
                    )}
                }
            }
        }
    }
    
    fun getCompletionRate(): Float {
        val state = _uiState.value
        return if (state.totalTargetMinutes > 0) {
            state.totalActualMinutes.toFloat() / state.totalTargetMinutes.toFloat()
        } else 0f
    }
    
    fun clearError() {
        _uiState.update { it.copy(error = null) }
    }
}

File: ios/StudyApp/Presentation/Plans/PlanViewModel.swift
import Combine
import Foundation

@MainActor
final class PlanViewModel: ScreenViewModel {
    @Published private(set) var plans: [StudyPlan] = []
    @Published private(set) var activePlan: StudyPlan?
    @Published private(set) var planItems: [PlanItem] = []
    @Published private(set) var subjects: [Subject] = []
    @Published var selectedDay: StudyWeekday?

    var weeklySchedule: [StudyWeekday: [PlanItemWithSubject]] {
        let subjectMap = Dictionary(uniqueKeysWithValues: subjects.map { ($0.id, $0) })
        return Dictionary(uniqueKeysWithValues: StudyWeekday.allCases.map { day in
            let dayItems = planItems
                .filter { $0.dayOfWeek == day }
                .compactMap { item -> PlanItemWithSubject? in
                    guard let subject = subjectMap[item.subjectId] else { return nil }
                    return PlanItemWithSubject(item: item, subject: subject)
                }
            return (day, dayItems)
        })
    }

    var totalTargetMinutes: Int {
        planItems.reduce(0) { $0 + $1.targetMinutes }
    }

    var completionRate: Double {
        guard totalTargetMinutes > 0 else { return 0 }
        let totalActual = planItems.reduce(0) { $0 + $1.actualMinutes }
        return min(Double(totalActual) / Double(totalTargetMinutes), 1)
    }

    func load() async {
        do {
            async let plansTask = app.persistence.getAllPlans()
            async let subjectsTask = app.persistence.getAllSubjects()
            let loadedPlans = try await plansTask
            plans = loadedPlans
            activePlan = loadedPlans.first(where: \.isActive)
            subjects = try await subjectsTask
            if let activePlan {
                planItems = try await app.persistence.getPlanItems(planId: activePlan.id)
                if selectedDay == nil {
                    selectedDay = StudyWeekday.allCases.first(where: { !(weeklySchedule[$0] ?? []).isEmpty }) ?? .monday
                }
            } else {
                planItems = []
                selectedDay = nil
            }
        } catch {
            app.present(error)
        }
    }

    func createPlan(name: String, startDate: Date, endDate: Date, items: [PlanItem]) {
        perform {
            let useCase = ManagePlansUseCase(repository: self.app.persistence)
            let syncedItems = items.map { item -> PlanItem in
                var value = item
                value.subjectSyncId = self.subjects.first(where: { $0.id == item.subjectId })?.syncId
                return value
            }
            try await useCase.createPlan(name: name, startDate: startDate, endDate: endDate, items: syncedItems)
            await self.load()
            self.app.bumpDataVersion()
        }
    }

    func savePlanItem(_ item: PlanItem) {
        perform {
            guard item.targetMinutes > 0 else { throw ValidationError(message: "目標時間は0より大きくしてください") }
            let subjectSyncId = self.subjects.first(where: { $0.id == item.subjectId })?.syncId
            let planSyncId = item.planSyncId ?? self.activePlan?.syncId
            if item.id == 0 {
                guard let activePlan = self.activePlan else {
                    throw ValidationError(message: "アクティブなプランがありません")
                }
                _ = try await self.app.persistence.insertPlanItem(
                    PlanItem(
                        planId: activePlan.id,
                        planSyncId: activePlan.syncId,
                        subjectId: item.subjectId,
                        subjectSyncId: subjectSyncId,
                        dayOfWeek: item.dayOfWeek,
                        targetMinutes: item.targetMinutes,
                        actualMinutes: item.actualMinutes,
                        timeSlot: item.timeSlot
                    )
                )
            } else {
                var updatedItem = item
                updatedItem.planSyncId = planSyncId
                updatedItem.subjectSyncId = subjectSyncId
                try await self.app.persistence.updatePlanItem(updatedItem)
            }
            await self.load()
            self.app.bumpDataVersion()
        }
    }

    func deletePlanItem(_ item: PlanItem) {
        perform {
            try await self.app.persistence.deletePlanItem(item)
            await self.load()
            self.app.bumpDataVersion()
        }
    }

    func deleteActivePlan() {
        perform {
            guard let activePlan = self.activePlan else { return }
            try await self.app.persistence.deletePlan(activePlan)
            await self.load()
            self.app.bumpDataVersion()
        }
    }
}
